src/marks/raster.js

```diff
@@ -0,0 +1,427 @@
+    case "nearest":
+      return rasterizeNearest;
+    case "barycentric":
+      return rasterizeBarycentric;
+    case "walk-on-spheres":
+      return rasterizeWalkOnSpheres;
+  }
+  throw new Error(`invalid rasterize: ${rasterize}`);
+}
+
+// Applies a simple forward mapping of samples, binning them into pixels without
+// any blending or interpolation.
+function rasterizeNone(canvas, index, {color}, {fill: F, fillOpacity: FO}, {x: X, y: Y}) {
+  const {width, height} = canvas;
+  const context2d = canvas.getContext("2d");
+  const image = context2d.createImageData(width, height);
+  const imageData = image.data;
+  let {r, g, b} = rgb(this.fill) ?? {r: 0, g: 0, b: 0};
+  let a = (this.fillOpacity ?? 1) * 255;
+  for (const i of index) {
+    if (X[i] < 0 || X[i] >= width || Y[i] < 0 || Y[i] >= height) continue;
+    const j = (Math.floor(Y[i]) * width + Math.floor(X[i])) << 2;
+    if (F) ({r, g, b} = rgb(color(F[i])));
+    if (FO) a = FO[i] * 255;
+    imageData[j + 0] = r;
+    imageData[j + 1] = g;
+    imageData[j + 2] = b;
+    imageData[j + 3] = a;
+  }
+  context2d.putImageData(image, 0, 0);
+}
+
+// Assumes that the fill and/or fillOpacity channels are a dense grid of values
+// that correspond to the size of the given canvas, in row-major order.
+function rasterizeDense(canvas, index, {color}, {fill: F, fillOpacity: FO}) {
+  const {width, height} = canvas;
+  const context2d = canvas.getContext("2d");
+  const image = context2d.createImageData(width, height);
+  const imageData = image.data;
+  let {r, g, b} = rgb(this.fill) ?? {r: 0, g: 0, b: 0};
+  let a = (this.fillOpacity ?? 1) * 255;
+  for (let i = 0, n = width * height; i < n; ++i) {
+    const j = i << 2;
+    if (F) {
+      const fi = F[i];
+      if (fi == null) {
+        imageData[j + 3] = 0;
+        continue;
+      }
+      ({r, g, b} = rgb(color(fi)));
+    }
+    if (FO) a = FO[i] * 255;
+    imageData[j + 0] = r;
+    imageData[j + 1] = g;
+    imageData[j + 2] = b;
+    imageData[j + 3] = a;
+  }
+  context2d.putImageData(image, 0, 0);
+}
+
+function rasterizeNearest(canvas, index, {color}, {fill: F, fillOpacity: FO}, {x: X, y: Y}) {
+  const {width, height} = canvas;
+  const context = canvas.getContext("2d");
+  const voronoi = Delaunay.from(
+    index,
+    (i) => X[i],
+    (i) => Y[i]
+  ).voronoi([0, 0, width, height]);
+  context.fillStyle = this.fill;
+  context.globalAlpha = this.fillOpacity;
+  for (let i = 0; i < index.length; ++i) {
+    context.beginPath();
+    voronoi.renderCell(i, context);
+    const j = index[i];
+    if (F) context.fillStyle = color(F[j]);
+    if (FO) context.globalAlpha = Math.abs(FO[j]);
+    context.fill();
+    if (context.globalAlpha === 1) (context.strokeStyle = context.fillStyle), context.stroke();
+  }
+}
+
+const ex = Symbol("extrapolate");
+
+function rasterizeBarycentric(canvas, index, {color}, {fill: F, fillOpacity: FO}, {x: X, y: Y}) {
+  const random = randomLcg(42); // TODO allow configurable rng?
```

Comment:
```suggestion
  const random = randomLcg(42);
```

Not sure how it would be useful. In fact I'd even want the rng to be reset across facets (when we have facets), to give more consistency.